src/Microsoft.AspNetCore.Blazor/Components/TreeParameterState.cs

```diff
@@ -0,0 +1,129 @@
+            // Defer instantiation of the result list until we know there's at least one
+            List<TreeParameterState> resultStates = null;
+
+            var numInfos = infos.Length;
+            for (var infoIndex = 0; infoIndex < numInfos; infoIndex++)
+            {
+                ref var info = ref infos[infoIndex];
+                if (TryGetMatchingProvider(info, componentState, out var provider))
+                {
+                    if (resultStates == null)
+                    {
+                        // Although not all parameters might be matched, we know the maximum number
+                        resultStates = new List<TreeParameterState>(infos.Length - infoIndex);
+                    }
+
+                    resultStates.Add(new TreeParameterState(info.LocalName, provider));
+                }
+            }
+
+            return resultStates;
+        }
+
+        private static bool TryGetMatchingProvider(in ReflectedTreeParameterInfo info, ComponentState componentState, out ProviderBase provider)
+        {
+            do
+            {
+                if (componentState.Component is ProviderBase candidateProvider
+                    && candidateProvider.CanSupplyValue(info.ValueType, info.ProviderName))
+                {
+                    provider = candidateProvider;
+                    return true;
+                }
+
+                componentState = componentState.ParentComponentState;
+            } while (componentState != null);
+
+            // No match
+            provider = null;
+            return false;
+        }
+
+        private static ReflectedTreeParameterInfo[] GetReflectedTreeParameterInfos(Type componentType)
+        {
+            if (!_cachedInfos.TryGetValue(componentType, out var infos))
+            {
+                infos = CreateReflectedTreeParameterInfos(componentType);
+                _cachedInfos[componentType] = infos;
+            }
+
+            return infos;
+        }
+
+        private static ReflectedTreeParameterInfo[] CreateReflectedTreeParameterInfos(Type componentType)
+        {
+            List<ReflectedTreeParameterInfo> result = null;
+            var candidateProps = ParameterCollectionExtensions.GetBindableProperties(componentType);
+            foreach (var prop in candidateProps)
+            {
+                var parameterAttribute = prop.GetCustomAttribute<ParameterAttribute>();
+                if (parameterAttribute.FromTree)
+                {
+                    if (result == null)
+                    {
+                        result = new List<ReflectedTreeParameterInfo>();
+                    }
+
+                    result.Add(new ReflectedTreeParameterInfo(
+                        prop.Name,
+                        prop.PropertyType,
+                        parameterAttribute.ProviderName));
+                }
+            }
+
+            return result?.ToArray();
```

Comment:
consider using using `Array.Empty<>` as a sentinel, then you don't need null checks elsewhere, and it's free.

Reply:
I've just tried it, but TBH I don't prefer it. It does remove one null check from elsewhere, but only if you're reading enough of the code to realise why the value can't be null. It doesn't help with perf because it's replacing a null check with executing (effectively) an array length check which is presumably about the same cost. If we had non-nullable types it would be clearer, but since we don't I'll leave as-is unless you feel strongly (in which case I am happy to change it really).

Reply:
I don't think it matters. I expect implementation details like this to require a detailed knowledge of what's going on.